src/transformers/integrations/flash_attention.py

```diff
@@ -0,0 +1,52 @@
+from typing import Optional, Tuple
+
+import torch
+
+from ..modeling_flash_attention_utils import _flash_attention_forward
+from ..utils import is_flash_attn_greater_or_equal_2_10
+
+
+_use_top_left_mask = not is_flash_attn_greater_or_equal_2_10()
+
+
+def flash_attention_forward(
+    module: torch.nn.Module,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attention_mask: Optional[torch.Tensor],
+    dropout: float = 0.0,
+    scaling: Optional[float] = None,
+    sliding_window: Optional[int] = None,
+    softcap: Optional[float] = None,
+    **kwargs,
+) -> Tuple[torch.Tensor, None]:
+    # This is before the transpose
+    seq_len = query.shape[2]
+
+    # FA2 uses non-transposed inputs
+    query = query.transpose(1, 2)
+    key = key.transpose(1, 2)
+    value = value.transpose(1, 2)
+
+    if query.dtype == torch.float32:
+        query = query.to(torch.float16)
+        key = key.to(torch.float16)
+        value = value.to(torch.float16)
```

Comment:
Any reason we specifically force float16? This check has been moved to the fa utils when providing a target dtype - seems like we have doubled semantics, imo it should be either but not both.

Reply:
Are all rope implementations fp32 agnostic? Iirc there has been a recast of qk to v dtype.

Reply:
I think this is more a typo and as you say we gotta default to flash attention's call!